src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -1675,7 +1675,10 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> Visitor<'l> for DumpVisitor<'l, 'tc
                         }
                     }
                     ty::TyTuple(..) => {}
-                    _ => span_bug!(ex.span, "Expected struct or tuple type, found {:?}", ty),
+                    _ => {
```

Comment:
When you're using the invalid tuple constructor is this ty:err or something similarly specific? Could we explicitly catch that case and return, and still `span_bug` ICE if it's something else? If we silently fail and cope then it's really hard to notice the omission until somebody reports a hard to discover bug against the RLS.

Reply:
It seems that almost all type variants (all but `TyInfer`?) can be crafted.

```rust
fn foo<T>(x: T) {
    bar.0; // TyError
    (0).0; // TyError
    x.0; // TyParam
    static S: i32 = 0;
    (S).0; // TyInt
    [].0; // TyArray
    extern { type Externed; }
    (|| -> Externed { panic!() })().0; // TyForeign
    (|| -> ! { panic!() })().0; // TyNever
    ...
}
```